apps/managedidentity/managedidentity.go

```diff
@@ -173,17 +280,104 @@ func createIMDSAuthRequest(ctx context.Context, id ID, resource string, claims s
 	return req, nil
 }
 
-func (client Client) getTokenForRequest(req *http.Request) (accesstokens.TokenResponse, error) {
+func createAzureArcAuthRequest(ctx context.Context, id ID, resource string, claims string) (*http.Request, error) {
+	identityEndpoint, _ := getAzureArcEnvironmentVariables()
+	var msiEndpoint *url.URL
+
+	if identityEndpoint == "" {
+		identityEndpoint = azureArcEndpoint
+	}
+
+	if _, ok := id.(systemAssignedValue); !ok {
+		return nil, errors.New("Azure Arc doesn't support specifying a user-assigned managed identity at runtime")
+	}
+
+	identityEndpoint = strings.Replace(identityEndpoint, "localhost", "127.0.0.1", -1)
+	msiEndpoint, parseErr := url.Parse(identityEndpoint)
+	if parseErr != nil {
+		return nil, fmt.Errorf("couldn't parse %q: %s", identityEndpoint, parseErr)
+	}
+
+	msiParameters := msiEndpoint.Query()
+	msiParameters.Set(apiVersionQueryParameterName, azureArcAPIVersion)
+	resource = strings.TrimSuffix(resource, "/.default")
+	msiParameters.Set(resourceQueryParameterName, resource)
+
+	if len(claims) > 0 {
+		msiParameters.Set("claims", claims)
+	}
+
+	msiEndpoint.RawQuery = msiParameters.Encode()
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, msiEndpoint.String(), nil)
+	if err != nil {
+		return nil, fmt.Errorf("error creating http request %s", err)
+	}
+	req.Header.Set(metaHTTPHeaderName, "true")
+	return req, nil
+}
+
+// GetEnvironmentVariables returns the identity and IMDS endpoints
+func getAzureArcEnvironmentVariables() (string, string) {
+	identityEndpoint := os.Getenv(IdentityEndpointEnvVar)
+	imdsEndpoint := os.Getenv(ArcIMDSEnvVar)
+
+	if identityEndpoint == "" || imdsEndpoint == "" {
+		platform := runtime.GOOS
+		fileDetectionPath, retrievedPath := azureArcFileDetection[platform]
+
+		if retrievedPath {
+			if _, err := os.Stat(fileDetectionPath); err == nil {
+				identityEndpoint = azureArcEndpoint
+				imdsEndpoint = himdsExecutableHelperString
+			}
+		}
+	}
+	return identityEndpoint, imdsEndpoint
+}
+
+func validateAzureArcEnvironment(identityEndpoint, imdsEndpoint string, platform string) bool {
```

Comment:
Some of this logic is pretty complex and we're going to struggle with supportability since there are no logs. Let's please add a story to have some logging, which we can tackle for GA. 

CC @chlowell for an opinion

Reply:
Created here: https://identitydivision.visualstudio.com/Engineering/_workitems/edit/3066894

Reply:
Logging needs a thoughtful end-to-end design because users, particularly Azure SDK, need to control MSAL's log output. I probably wouldn't block GA on it because MSAL doesn't log anything today and managed identity clients will return a string from the point of failure when anything goes wrong.